### 1. Równanie normalne:

![image.png](równanie_normalne.png)

In [1]:
import numpy as np

### 2. Przygotowanie danych

In [2]:
X1 = np.array([1, 2, 3, 4, 5, 6])
Y = np.array([3000, 3250, 3500, 3750, 4000, 4250])
m = len(X1)

print(f'Lata pracy: {X1}')
print(f'Wynagrodzenie: {Y}')
print(f'Liczba próbek: {m}')

Lata pracy: [1 2 3 4 5 6]
Wynagrodzenie: [3000 3250 3500 3750 4000 4250]
Liczba próbek: 6


#### - reshape - zmiana kształtu danych na 6 wierszy i 1 kolumnę

In [9]:
X1 = X1.reshape(m,1)
print(X1)

[[1]
 [2]
 [3]
 [4]
 [5]
 [6]]


#### - bias - przygotowanie wektoru jedynek

In [14]:
bias = np.ones((m,1))
print(bias)
print(bias.shape)

[[1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]
(6, 1)


#### - połaczenie bias i zmiennej X1

In [19]:
X = np.append(bias, X1,axis=1)
print(X)
print(X.shape)

[[1. 1.]
 [1. 2.]
 [1. 3.]
 [1. 4.]
 [1. 5.]
 [1. 6.]]
(6, 2)


### <a name='2'></a> 3. Równanie normalne - numpy

Regresja liniowa w $R^2$:
$$Y = w_0 + w_1X_1$$
$$Y = W^TX$$
gdzie: $$W= \left[\begin{matrix}w_0\\w_1\end{matrix}\right]$$ $$ X= \left[\begin{matrix}1\\X_1\end{matrix}\right] $$ stąd $$ W^T= \left[\begin{matrix}w_0&w_1\end{matrix}\right] $$    
$$Y = W^TX = \left[\begin{matrix}w_0&w_1\end{matrix}\right] \cdot  \left[\begin{matrix}1\\X_1\end{matrix}\right] = w_0 + w_1X_1$$




Równanie normalne - równanie pozwalające obliczyć minimum funkcji straty (o ile istnieje)

$$W = (X^TX)^{-1}(X^TY)$$

In [29]:
X

array([[1., 1.],
       [1., 2.],
       [1., 3.],
       [1., 4.],
       [1., 5.],
       [1., 6.]])

In [24]:
# transponowanie X

X.T

array([[1., 1., 1., 1., 1., 1.],
       [1., 2., 3., 4., 5., 6.]])

In [25]:
# mnożenie macierzy

np.dot(X.T,X)

array([[ 6., 21.],
       [21., 91.]])

In [37]:
# obliczenie lewej strony równania - (XT*X)-1 - obliczanie macierzy odwrotnej (podniesienie do potęgi - 1)

L = np.linalg.inv(np.dot(X.T,X))
L

array([[ 0.86666667, -0.2       ],
       [-0.2       ,  0.05714286]])

In [38]:
# obliczanie prwej strony - transponowane X * Y
P = np.dot(X.T, Y)
P

array([21750., 80500.])

In [40]:
# finałowe mnożenie macierzy - lewej i prawej strony
W = np.dot(L,P)
W

array([2750.,  250.])

#### - 2750 - w0(przecięcie), w1 (nachylenie) 250


Końcowa postać modelu$$Y = 2750 + 250X_1$$

### 4. Regresja liniowa scikit learn - jak możemy dojść do wag

In [43]:
# tutaj nie musimy łączyć macierzu 1 z X1, sklearn robi to autmatycznie

from sklearn.linear_model import LinearRegression

regression = LinearRegression()
regression.fit(X1, Y)

print(regression.intercept_) # w0 - przecięcie
print(regression.coef_[0]) # coefficient - współczynnik

2750.0
250.0
